File: packages/js/sdk/src/wrap-agent.ts
import { HumanMessage } from '@langchain/core/messages';
import type { BaseMessage } from '@langchain/core/messages';
import { BaseCallbackHandler } from '@langchain/core/callbacks/base';
import type { Serialized } from '@langchain/core/load/serializable';
import type { TrikAgent, TrikContext, TrikResponse, TrikProgressEvent } from '@trikhub/manifest';
import { extractToolInfo } from './interceptor.js';
import { getActiveWorkspaceToolNames } from './workspace-tools.js';
import { getActiveRegistryToolNames } from './registry-tools.js';

/**
 * Any agent with a LangGraph-compatible invoke method.
 * Works with createReactAgent, custom StateGraphs, or any object implementing invoke().
 */
export interface InvokableAgent {
  invoke(
    input: { messages: BaseMessage[] },
    config?: unknown
  ): Promise<{ messages: BaseMessage[] }>;
}

/**
 * Factory function that creates an agent from TrikContext.
 * Use this when your agent needs config values (e.g., API keys) at creation time.
 */
export type AgentFactory = (
  context: TrikContext
) => InvokableAgent | Promise<InvokableAgent>;

export interface WrapAgentOptions {
  // Reserved for future options (domain prompting, etc.)
}

function isInvokable(obj: unknown): obj is InvokableAgent {
  return (
    typeof obj === 'object' &&
    obj !== null &&
    'invoke' in obj &&
    typeof (obj as Record<string, unknown>).invoke === 'function'
  );
}

/**
 * LangGraph callback handler that bridges tool execution events
 * to the TrikContext.onProgress callback.
 */
class ProgressCallbackHandler extends BaseCallbackHandler {
  name = 'TrikProgressHandler';
  private activeTools = new Map<string, string>();

  constructor(private onProgress: (event: TrikProgressEvent) => void) {
    super();
  }

  handleToolStart(
    tool: Serialized,
    _input: string,
    runId: string,
    _parentRunId?: string,
    _tags?: string[],
    _metadata?: Record<string, unknown>,
    runName?: string,
  ): void {
    const toolName = runName || ('id' in tool ? tool.id[tool.id.length - 1] : 'unknown');
    this.activeTools.set(runId, toolName);
    this.onProgress({ type: 'tool_start', toolName });
  }

  handleToolEnd(output: unknown, runId: string): void {
    const toolName = this.activeTools.get(runId) || '';
    this.activeTools.delete(runId);
    this.onProgress({ type: 'tool_end', toolName });
  }

  handleToolError(_err: Error, runId: string): void {
    const toolName = this.activeTools.get(runId) || '';
    this.activeTools.delete(runId);
    this.onProgress({ type: 'tool_error', toolName });
  }
}

/**
 * Wrap a LangGraph agent (or factory) into a TrikAgent.
 *
 * Handles:
 * 1. Message history management across turns within a session
 * 2. Tool call extraction from LangGraph messages → ToolCallRecord[]
 * 3. transfer_back detection → sets transferBack flag
 *
 * @param agentOrFactory - A pre-built agent with invoke(), or a factory (context) => agent
 * @param options - Reserved for future use
 *
 * @example
 * // Pattern 1: Pre-built agent
 * const agent = createReactAgent({ llm, tools: [...myTools, transferBackTool] });
 * export default wrapAgent(agent);
 *
 * @example
 * // Pattern 2: Factory (when you need config at runtime)
 * export default wrapAgent(async (context) => {
 *   const llm = new ChatAnthropic({ apiKey: context.config.get("ANTHROPIC_API_KEY") });
 *   return createReactAgent({ llm, tools: [...myTools, transferBackTool] });
 * });
 */
export function wrapAgent(
  agentOrFactory: InvokableAgent | AgentFactory,
  _options?: WrapAgentOptions
): TrikAgent {
  let resolvedAgent: InvokableAgent | null = isInvokable(agentOrFactory)
    ? agentOrFactory
    : null;

  // Per-session message history
  const sessionMessages = new Map<string, BaseMessage[]>();

  return {
    async processMessage(
      message: string,
      context: TrikContext
    ): Promise<TrikResponse> {
      // Lazy init for factory pattern (created once, reused across sessions)
      if (!resolvedAgent) {
        resolvedAgent = await (agentOrFactory as AgentFactory)(context);
      }

      // Determine which internal tools are active (for output filtering)
      const workspaceToolNames = getActiveWorkspaceToolNames(context.capabilities);
      const registryToolNames = getActiveRegistryToolNames(context.capabilities);

      // Get or create session message history
      let messages = sessionMessages.get(context.sessionId);
      if (!messages) {
        messages = [];
        sessionMessages.set(context.sessionId, messages);
      }

      // Record where new messages start (for extracting this turn's tool calls)
      const startIndex = messages.length;

      // Add user message
      messages.push(new HumanMessage(message));

      // Invoke the LangGraph agent (with progress callbacks if available)
      const config = context.onProgress
        ? { callbacks: [new ProgressCallbackHandler(context.onProgress)] }
        : undefined;
      const result = await resolvedAgent.invoke({ messages }, config);

      // Store the full updated message history
      sessionMessages.set(context.sessionId, result.messages);

      // Extract tool calls and transfer-back signal from new messages only
      const { toolCalls, transferBack, responseMessage } = extractToolInfo(
        result.messages,
        startIndex
      );

      // Filter out internal tool calls from the output (workspace + registry)
      const internalToolNames = new Set([...workspaceToolNames, ...registryToolNames]);
      const filteredToolCalls = internalToolNames.size > 0
        ? toolCalls.filter((tc) => !internalToolNames.has(tc.tool))
        : toolCalls;

      return {
        message: responseMessage,
        transferBack,
        toolCalls: filteredToolCalls.length > 0 ? filteredToolCalls : undefined,
      };
    },
  };
}

File: packages/python/trikhub/sdk/wrap_agent.py
"""
Wrap a LangGraph agent into a TrikAgent with processMessage support.

Maintains per-session message history, lazy-initializes factory agents,
and extracts tool calls + transfer_back signals from the LangGraph response.
Mirrors packages/js/sdk/src/wrap-agent.ts.
"""

from __future__ import annotations

from typing import Any, Protocol, Union, runtime_checkable

from langchain_core.messages import BaseMessage, HumanMessage, SystemMessage

from trikhub.manifest import TrikContext, TrikResponse

from .interceptor import extract_tool_info
from .workspace_tools import WORKSPACE_SYSTEM_PROMPT, get_active_workspace_tool_names
from .registry_tools import REGISTRY_SYSTEM_PROMPT, get_active_registry_tool_names


@runtime_checkable
class InvokableAgent(Protocol):
    """A LangGraph agent that can be invoked with messages."""

    async def ainvoke(
        self, input: dict[str, Any], config: Any = None
    ) -> dict[str, Any]: ...


AgentFactory = Union[
    type["_Callable_AgentFactory"],
    Any,  # Callable[[TrikContext], InvokableAgent | Awaitable[InvokableAgent]]
]


class _Callable_AgentFactory(Protocol):
    """Protocol for agent factory callables."""

    def __call__(self, context: TrikContext) -> Any: ...


class _WrappedAgent:
    """TrikAgent implementation wrapping a LangGraph agent."""

    def __init__(self, agent_or_factory: Any) -> None:
        self._agent_or_factory = agent_or_factory
        self._resolved_agent: InvokableAgent | None = None
        self._session_messages: dict[str, list[BaseMessage]] = {}
        self._session_has_system_prompt: set[str] = set()

        # If it's already an invokable agent (has ainvoke), use it directly
        if hasattr(agent_or_factory, "ainvoke"):
            self._resolved_agent = agent_or_factory

    async def _get_agent(self, context: TrikContext) -> InvokableAgent:
        """Lazily resolve the agent from a factory if needed."""
        if self._resolved_agent is None:
            result = self._agent_or_factory(context)
            # Support async factories
            if hasattr(result, "__await__"):
                result = await result
            self._resolved_agent = result
        return self._resolved_agent

    async def process_message(self, message: str, context: TrikContext) -> TrikResponse:
        """Process a user message through the wrapped LangGraph agent.

        Maintains per-session message history and extracts tool calls
        from the agent's response. When capabilities are present,
        filters workspace tool calls from output and prepends system prompt.
        """
        agent = await self._get_agent(context)

        # Determine which internal tools are active (for output filtering)
        workspace_tool_names = get_active_workspace_tool_names(context.capabilities)
        registry_tool_names = get_active_registry_tool_names(context.capabilities)

        # Get or create session history
        session_id = context.sessionId
        messages = self._session_messages.get(session_id, [])

        # Prepend system prompts on first message of a session
        if session_id not in self._session_has_system_prompt:
            if workspace_tool_names:
                messages.append(SystemMessage(content=WORKSPACE_SYSTEM_PROMPT))
            if registry_tool_names:
                messages.append(SystemMessage(content=REGISTRY_SYSTEM_PROMPT))
            if workspace_tool_names or registry_tool_names:
                self._session_has_system_prompt.add(session_id)

        # Mark start of this turn for interceptor
        start_index = len(messages)

        # Append user message
        messages.append(HumanMessage(content=message))

        # Invoke the agent
        result = await agent.ainvoke({"messages": messages})

        # Update session history with full result
        result_messages: list[BaseMessage] = result.get("messages", messages)
        self._session_messages[session_id] = result_messages

        # Extract tool info from new messages only
        info = extract_tool_info(result_messages, start_index)

        # Filter out internal tool calls from the output (workspace + registry)
        internal_tool_names = workspace_tool_names | registry_tool_names
        if internal_tool_names:
            filtered_tool_calls = [
                tc for tc in info.tool_calls if tc.tool not in internal_tool_names
            ]
        else:
            filtered_tool_calls = info.tool_calls

        return TrikResponse(
            message=info.response_message,
            transferBack=info.transfer_back,
            toolCalls=filtered_tool_calls if filtered_tool_calls else None,
        )


def wrap_agent(agent_or_factory: Any) -> _WrappedAgent:
    """Wrap a LangGraph agent (or factory) into a TrikAgent.

    The agent can be:
    - A pre-built LangGraph agent with an `ainvoke` method
    - A factory function `(context: TrikContext) -> agent` (sync or async)
      The factory is called once lazily on first message, then cached.

    The returned object implements the TrikAgent protocol with `process_message()`.

    Args:
        agent_or_factory: A LangGraph agent instance or factory function

    Returns:
        A TrikAgent-compatible object with process_message()
    """
    return _WrappedAgent(agent_or_factory)
